### PR TITLE
build(deps): bump react-native from 0.72.17 to 0.73.9 #WIP #DONOTMERGE

### PR DESCRIPTION
**Pull request by gkartalis**
This PR resolves [PHIRE-1181] 

### Description

Bumps react-native version from `0.72.17` to `0.73.9`

#### What changed:

- Debugging improvements
- Console Log history in Hermes
- Updated debugging [docs](https://reactnative.dev/docs/0.73/debugging)
- Experimental New Debugger
- Stable Symlink Support in Metro
- Better Android 14 support
- Java bump from 11 to 17
- Gradle Bump from 7.4.x to 8.1.x

You can read more info on [RN 73 official blogpost](https://reactnative.dev/blog/2023/12/06/0.73-debugging-improvements-stable-symlinks)

#### Which Deps were bumped and why:

- `@braze/react-native-sdk`: `5.2.0` ~> `8.4.0`
- `@react-native-async-storage/async-storage`: `1.19.3` ~> `1.19.8`
- `@react-native-clipboard/clipboard`: `1.7.0` ~> `1.14.1`
- `react-native-code-push`: `8.1.0` ~> `8.3.1`
- `react-native-gesture-handler`: `2.13.1` ~> `2.19.0`
- `react-native-webview`: `13.6.2` ~> `13.6.3`

React Native 0.73 will depend on Android Gradle Plugin (AGP) 8.x. This will require all the libraries to specify a namespace in their build.gradle file. RN team added a compatibility layer for libraries that haven't specified a namespace, but please consider updating your libraries nonetheless.

Read more [here](https://github.com/react-native-community/discussions-and-proposals/issues/671)

#### Patches:

Enhanced react native patch with `.mock('../Libraries/Image/Image', () => mockComponent('../Libraries/Image/Image'))` due to [this issue](https://github.com/facebook/react-native/issues/41907#issuecomment-2293385777).

#### DevExperience Before / After:

Metro is way faster when bundling the app (around ~2x faster)

Android builds have an insane boost that is mostly due to java 17 and gradle updates:

On emulator with cleared cache / deleted app:

|Before|After|
|---|---|
|androidBuildBefore|Screenshot 2024-09-13 at 12 43 39|

on emulator with cached build:

|Before|After|
|---|---|
|Screenshot 2024-09-13 at 13 22 59|Screenshot 2024-09-13 at 12 44 32|

#### Steps To Upgrade on your machine after this gets merged:

```sh
# Delete node modules and caches:
rm -rf node_modules && rm -rf .cache && rm -rf "$TMPDIR/metro*" && rm -rf "$TMPDIR/haste-map-*"

# install all the dependencies needed
yarn install:all

# you will most likely have to run also the following to bump hermes-engine:
cd ios && bundle exec pod update hermes-engine --no-repo-update & cd ..
```

Afterwards in order to bump JAVA:

For asdf users:

```sh
asdf install
```

For non-asdf users:

```sh
brew install --cask zulu@17

# Get path to where cask was installed to double-click installer
brew info --cask zulu@17
```

> [!IMPORTANT]  
> For all (asdf/ non asdf users). After you install the JDK, update your JAVA_HOME environment variable in your `.zshrc` file. Afterwards run `source ~/.zshrc` to make it effective and restart your terminal.
>If you used above steps, JDK will likely be at /Library/Java/JavaVirtualMachines/zulu-17.jdk/Contents/Home.
> You can try out in the terminal by typing $JAVA_HOME to make sure that it was updated properly.

 
### PR Checklist

- [ ] I have tested my changes on **iOS** and **Android**.
- [x] I hid my changes behind a **[feature flag]**, or they don't need one.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [x] I added an **[app state migration]**, or my changes do not require one.
- [ ] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [x] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

-

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

- bumps react-native to 0.73.9 - gkartalis

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md

[PHIRE-1181]: https://artsyproduct.atlassian.net/browse/PHIRE-1181?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .tool-versions [hunk 1]
```diff
@@ -1,3 +1,3 @@
 nodejs 20.9.0
 ruby 3.1.4
-java zulu-11.58.15
+java zulu-17.50.19
```
Comment (gkartalis) [inline .tool-versions:3]: Java bump to 17

File: android/app/src/main/java/net/artsy/app/MainActivity.kt [hunk 2]
```diff
@@ -0,0 +1,99 @@
+package net.artsy.app
```
Comment (gkartalis) [inline android/app/src/main/java/net/artsy/app/MainActivity.kt:1]: we had to re-write this with Kotlin 🙈

File: android/app/src/main/java/net/artsy/app/MainApplication.kt [hunk 3]
```diff
@@ -0,0 +1,86 @@
+package net.artsy.app
+
```
Comment (gkartalis) [inline android/app/src/main/java/net/artsy/app/MainApplication.kt:2]: we had to re-write this with Kotlin 🙈

**Comment by ArtsyOpenSource** [comment]
### This PR contains the following changes:

- Dev changes (bumps react-native to 0.73.9 - gkartalis)
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 21b596ed069429947aa76b84d2d02a3e835a035c
</p>

File: metro.config.js [hunk 4]
```diff
@@ -19,6 +20,9 @@ const config = {
 
   resolver: {
     resolverMainFields: ["sbmodern", "react-native", "browser", "main"], // needed for storybook
+    extraNodeModules: {
+      images: path.resolve(__dirname, "./images"), // Add this line for Metro to resolve 'images folder'
+    },
```
Comment (gkartalis) [inline metro.config.js:25]: Big GOTCHA, `images/` path stopped being recognised after the rn bump and we had to let metro config know where they are

File: package.json [hunk 5]
```diff
@@ -117,8 +117,8 @@
     "@invertase/react-native-apple-authentication": "2.1.5",
     "@kesha-antonov/react-native-action-cable": "1.1.4",
     "@ptomasroos/react-native-multi-slider": "2.2.2",
-    "@react-native-async-storage/async-storage": "1.19.3",
-    "@react-native-clipboard/clipboard": "1.7.0",
+    "@react-native-async-storage/async-storage": "1.19.8",
+    "@react-native-clipboard/clipboard": "1.14.1",
```
Comment (gkartalis) [inline package.json:121]: Had to bump these in order for them to be rn 73 compliant

File: package.json [hunk 6]
```diff
@@ -155,19 +155,20 @@
     "query-string": "4.3.4",
     "react": "18.2.0",
     "react-fps": "1.0.6",
-    "react-native": "0.72.17",
+    "react-native": "0.73.9",
     "react-native-blob-util": "0.19.9",
     "react-native-blurhash": "1.1.11",
     "react-native-bootsplash": "3.2.0",
-    "react-native-code-push": "8.1.0",
+    "react-native-code-push": "8.3.1",
```
Comment (gkartalis) [inline package.json:162]: Had to bump in order for them to be rn 73 compliant

File: package.json [hunk 7]
```diff
@@ -190,7 +191,7 @@
     "react-native-url-polyfill": "2.0.0",
     "react-native-view-shot": "3.8.0",
     "react-native-vimeo-iframe": "1.2.1",
-    "react-native-webview": "13.6.2",
+    "react-native-webview": "13.6.3",
```
Comment (gkartalis) [inline package.json:193]: Had to bump in order for them to be rn 73 compliant

File: package.json [hunk 8]
```diff
@@ -155,19 +155,20 @@
...
     "react-native-config": "https://github.com/artsy/react-native-config.git#v1.4.12-artsy",
     "react-native-context-menu-view": "git+https://github.com/artsy/react-native-context-menu-view.git#v1.10.10-artsy",
     "react-native-dev-menu-android": "1.0.10",
     "react-native-device-info": "10.3.0",
     "react-native-document-picker": "9.3.0",
     "react-native-fast-image": "8.6.3",
     "react-native-fbsdk-next": "13.0.0",
-    "react-native-gesture-handler": "2.13.1",
+    "react-native-flipper": "0.265.0",
+    "react-native-gesture-handler": "2.19.0",
```
Comment (gkartalis) [inline package.json:170]: Had to bump in order for them to be rn 73 compliant

**Review by brainbicycle — CHANGES_REQUESTED** [review]
This looks great! 
Two small ones:
- the braze push handling in MainActivity is commented out, I think it is still necessary? 🤔 
- I think the react native patch may have gotten some generated code carried along

File: scripts/ci/build-for-tests-ios [hunk 9]
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 set -euxo pipefail
 
-xcodebuild -workspace "$WORKSPACE" -scheme "$SCHEME" -configuration Debug -sdk iphonesimulator build -destination platform="$DEVICE_HOST_PLAT",OS="$DEVICE_HOST_OS",name="$DEVICE_HOST_NAME" -derivedDataPath "$DERIVED_DATA_PATH" GCC_PREPROCESSOR_DEFINITIONS='$(inherited) CI_DISABLE_FLIPPER=1' |
+xcodebuild -workspace "$WORKSPACE" -scheme "$SCHEME" -configuration Debug -sdk iphonesimulator build -destination platform="$DEVICE_HOST_PLAT",OS="$DEVICE_HOST_OS",name="$DEVICE_HOST_NAME" -derivedDataPath "$DERIVED_DATA_PATH" |
```
Comment (brainbicycle) [inline scripts/ci/build-for-tests-ios:4]: here

File: HACKS.md [hunk 10]
```diff
@@ -304,6 +314,17 @@ patch.
...
+## Patch-package for @braze/react-native-sdk
+
+#### When can we remove this:
+
+When we upgrade @braze/react-native-sdk to version >= 8.3.0.
+
+#### Explanation/Context:
+
+We had to patch Braze in order to proceed with the react native upgrade to 0.73.9. The patch was needed to support kotlin jvm target 17 and
+in order to also make targetCompatibility and sourceCompatibility compatible with the JAVA 17 which is the standard for newer rn versions starting 0.73.
```
[the remark this answers is not in the document]
Comment (gkartalis) [reply HACKS.md:326]: 💯

File: src/app/Components/PopoverMessage/PopoverMessageProvider.tsx [hunk 11]
```diff
@@ -22,7 +22,7 @@ export const PopoverMessageProvider: React.FC = ({ children }) => {
   const [popoverMessage, setPopoverMessage] = useState<PopoverMessageItem | null>(null)
   const showingPopoverMessage = useRef<boolean>(false)
   const lastStartedAt = useRef<number | null>(null)
-  const timer = useRef<number | null>(null)
```
[the remark this answers is not in the document]
Comment (gkartalis) [reply src/app/Components/PopoverMessage/PopoverMessageProvider.tsx:25]: it is weird indeed 🤷

File: android/app/src/main/java/net/artsy/app/MainApplication.kt [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+        val applicationContext = this
+        FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
+            if (!task.isSuccessful) {
+                // Log.i(TAG, "TOKEN failed to fetch registration token")
+                // Log.w(TAG, "Fetching FCM registration token failed", task.exception)
+                return@addOnCompleteListener
+            }
+            val token = task.result
+            // Log.i(TAG, "TOKEN firebase messaging token $token")
+            // Braze.getInstance(applicationContext).setRegisteredPushToken(token)
```
[the remark this answers is not in the document]
Comment (gkartalis) [reply android/app/src/main/java/net/artsy/app/MainApplication.kt]: I was getting an error here when trying to build ~ I think that in the end we will need to bump braze to be compliant 🫨 .

Seems like they moved this function to JS side ~ will trigger betas after I address that to test if it works

File: HACKS.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (gkartalis) [reply HACKS.md:326]: we will probably need to test out both android & iOS thoroughly, had to bump to 8.4.0 in the end - will remove the hack entry

**Comment by gkartalis** [comment]
Bumped braze to 8.4.0 in the end ~ moved the registration of the token to the JS side since I was getting build errors from a depracated method from `setRegisteredPushToken` tried some others and no luck thus I moved the token to the JS side.

Let me know if you have any concerns ~ but I Was thinking to open a separate PR for braze 🤷‍♂️

**Comment by gkartalis** [comment]
FreshBetas:

`android-8.51.0-1674645836`
`ios - 8.51.0 - 2024.9.13.15`

**Comment by gkartalis** [comment]
Braze works! Received notifications in both android / ios betas above

**Review by MrSltun — DISMISSED** [review]
Thank you for pushing through this!! 💪 ⚡ 📈

File: src/app/utils/PushNotification.ts [hunk 13]
```diff
@@ -235,6 +236,8 @@ export async function configure() {
     // (optional) Called when Token is generated (iOS and Android)
     onRegister: async (token) => {
       try {
+        registerPushToken(token.token)
```
Comment (brainbicycle) [inline src/app/utils/PushNotification.ts:239]: nice

**Review by brainbicycle — APPROVED** [review]

**Pull request merged**